internal/provisioner/objects/object.go

```diff
@@ -38,3 +45,25 @@ func NewUnprivilegedPodSecurity() *corev1.PodSecurityContext {
 		RunAsNonRoot: &nonRoot,
 	}
 }
+
+// objectGetter get the object with name
+type objectGetter func(ctx context.Context, cli client.Client, namespace, name string) (client.Object, error)
+
+// EnsureObjectDeleted ensures that a object with the name is deleted
+// if Contour owner labels exist.
+func EnsureObjectDeleted(ctx context.Context, cli client.Client, contour *model.Contour, name string, getter objectGetter) error {
+	obj, err := getter(ctx, cli, contour.Namespace, name)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			return nil
+		}
+		return err
+	}
+	if !labels.Exist(obj, model.OwnerLabels(contour)) {
+		return nil
+	}
+	if err := cli.Delete(ctx, obj); err == nil || errors.IsNotFound(err) {
+		return nil
+	}
+	return err
```

Comment:
It looks like this is returning the wrong `err` if there was an error deleting the object, i.e. the one declared on L55 rather than the one returned from the delete. Let's invert this to avoid such issues:

```suggestion
	if err := cli.Delete(ctx, obj); err != nil && !errors.IsNotFound(err) {
		return err
	}
	
	return nil
```

Reply:
oops, the L65's err is shadowed, maybe unshadowing it is better for read?